### PR TITLE
Dataset array, dframe and redim method fixes

### PR DESCRIPTION
**Pull request by philippjfr**
Various fixes for Dataset methods along with unit tests.

### CODE REVIEW

File: holoviews/element/comparison.py [hunk 1]
```diff
@@ -257,6 +262,11 @@ def compare_arrays(cls, arr1, arr2, msg='Arrays'):
                 raise cls.failureException(msg + str(e)[11:])
 
     @classmethod
+    def compare_dataframe(cls, df1, df2, msg='DataFrames'):
+        if not df1.equals(df2):
+            raise cls.failureException('%s are not equal' % msg)
+
```
Comment (jlstevens) [inline holoviews/element/comparison.py]: What happened before we added this? Surely we had some dataframes to compare in the notebook data tests?

Comment (philippjfr) [reply holoviews/element/comparison.py]: Nope I think we always avoided it somehow.

**Comment by philippjfr** [comment]
Ready to merge.

File: holoviews/core/data/xarray.py [hunk 2]
```diff
@@ -272,9 +272,9 @@ def length(cls, dataset):
     @classmethod
     def dframe(cls, dataset, dimensions):
         if dimensions:
-            return dataset.reindex(columns=dimensions).data.to_dataframe().reset_index(dimensions)
+            return dataset.data.to_dataframe().reset_index()[dimensions]
         else:
-            return dataset.data.to_dataframe().reset_index(dimensions)
+            return dataset.data.to_dataframe()
```
Comment (jlstevens) [inline holoviews/core/data/xarray.py]: Does this result in the same behavior as before (just written more clearly) or is the index no longer reset in the second branch?

File: tests/testdataset.py [hunk 3]
```diff
@@ -244,7 +244,7 @@ def test_dataset_2D_aggregate_partial_hm(self):
                           kdims=['x', 'y'], vdims=['z'])
         self.assertEqual(dataset.aggregate(['x'], np.mean),
                          Dataset({'x':self.xs, 'z':z_ints}, kdims=['x'], vdims=['z']))
-
+        
     # Indexing
```
Comment (jlstevens) [inline tests/testdataset.py]: Trailing whitespace!!

**Comment by jlstevens** [comment]
Looks good. Got a few questions in the comments above and please remove the trailing whitespace. :-)

File: holoviews/core/data/xarray.py (hunk 2 above)
Comment (philippjfr) [reply holoviews/core/data/xarray.py]: Thanks, fixed and added a unit test for that branch too.

**Comment by philippjfr** [comment]
Okay, ready this time.

**Comment by jlstevens** [comment]
I'll merge once the tests are green.

**Pull request merged**